src/tool/hpcrun/gpu/opencl/opencl-api.c

```diff
@@ -0,0 +1,461 @@
+//   this software without specific prior written permission.
+//
+// This software is provided by RICE and contributors "as is" and any
+// express or implied warranties, including, but not limited to, the
+// implied warranties of merchantability and fitness for a particular
+// purpose are disclaimed. In no event shall RICE or contributors be
+// liable for any direct, indirect, incidental, special, exemplary, or
+// consequential damages (including, but not limited to, procurement of
+// substitute goods or services; loss of use, data, or profits; or
+// business interruption) however caused and on any theory of liability,
+// whether in contract, strict liability, or tort (including negligence
+// or otherwise) arising in any way out of the use of this software, even
+// if advised of the possibility of such damage.
+//
+// ******************************************************* EndRiceCopyright *
+
+//******************************************************************************
+// system includes
+//******************************************************************************
+
+#include <assert.h>
+#define CL_TARGET_OPENCL_VERSION 120
+#include <CL/cl.h>
+#include <inttypes.h>
+
+
+
+//******************************************************************************
+// local includes
+//******************************************************************************
+
+#include <hpcrun/safe-sampling.h>
+#include <hpcrun/gpu/gpu-activity.h>
+#include <hpcrun/gpu/gpu-activity-channel.h>
+#include <hpcrun/gpu/gpu-activity-process.h>
+#include <hpcrun/gpu/gpu-correlation-channel.h>
+#include <hpcrun/gpu/gpu-correlation-id-map.h>
+#include <hpcrun/gpu/gpu-application-thread-api.h>
+#include <hpcrun/gpu/gpu-metrics.h>
+#include <hpcrun/gpu/gpu-monitoring-thread-api.h>
+#include <hpcrun/gpu/gpu-op-placeholders.h>
+#include <hpcrun/messages/messages.h>
+#include <hpcrun/sample-sources/libdl.h>
+#include <lib/prof-lean/stdatomic.h>
+#include <lib/prof-lean/usec_time.h>
+
+#include "opencl-api.h"
+#include "opencl-activity-translate.h"
+#include "opencl-intercept.h"
+#include "opencl-memory-manager.h"
+
+
+
+//******************************************************************************
+// macros
+//******************************************************************************
+
+#define CPU_NANOTIME() (usec_time() * 1000)
+
+#define opencl_path() "libOpenCL.so"
+
+#define FORALL_OPENCL_ROUTINES(macro)	      \
+  macro(clGetEventProfilingInfo)   		      \
+  macro(clReleaseEvent)                     \
+  macro(clSetEventCallback)
+
+#define OPENCL_FN_NAME(f) DYN_FN_NAME(f)
+
+#define OPENCL_FN(fn, args)                 \
+  cl_int (*OPENCL_FN_NAME(fn)) args
+
+#define HPCRUN_OPENCL_CALL(fn, args)        \
+{                                           \
+  cl_int status = OPENCL_FN_NAME(fn) args;  \
+  if (status != CL_SUCCESS) {               \
+    opencl_error_report(status);            \
```

Comment:
This does nothing. It returns an error string that is not printed. should an error cause the program to fail with an error?

Reply:
will this be ok?
```
if (status != CL_SUCCESS) {                                                 \
    ETMSG(OPENCL, "opencl call failed: %s", opencl_error_report(status));     \
  }
```

Reply:
this is OK.